plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataUpdater.java

```diff
@@ -0,0 +1,237 @@
+
+  private final Maven2MetadataMerger metadataMerger;
+
+  private final MetadataXpp3Reader metadataReader;
+
+  private final MetadataXpp3Writer metadataWriter;
+
+  private final DateFormat dotlessTimestampFormat;
+
+  private final DateFormat dottedTimestampFormat;
+
+  @Inject
+  public MetadataUpdater(final MavenFacet mavenFacet) {
+    this.mavenFacet = checkNotNull(mavenFacet);
+    this.metadataMerger = new Maven2MetadataMerger();
+    this.metadataReader = new MetadataXpp3Reader();
+    this.metadataWriter = new MetadataXpp3Writer();
+    this.dotlessTimestampFormat = new SimpleDateFormat("yyyyMMddHHmmss");
+    this.dotlessTimestampFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+    this.dottedTimestampFormat = new SimpleDateFormat("yyyyMMdd.HHmmss");
+    this.dottedTimestampFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+  }
+
+  /**
+   * Writes/updates metadata, merges existing one, if any.
+   */
+  public void update(final MavenPath mavenPath, final Maven2Metadata metadata) {
+    checkNotNull(mavenPath);
+    checkNotNull(metadata);
+    try {
+      final Metadata oldMetadata = read(mavenPath);
+      if (oldMetadata == null) {
+        // old does not exists, just write it
+        write(mavenPath, toMetadata(metadata));
+      }
+      else {
+        // TODO: compare? unsure is it worth it, as compare would also eat CPU maybe even more that writing would
+        // update old by merging them and write out
+        final Metadata updated = metadataMerger.merge(
+            ImmutableList.of(
+                new MetadataEnvelope("old:" + mavenPath.getPath(), oldMetadata),
+                new MetadataEnvelope("new" + mavenPath.getPath(), toMetadata(metadata))
+            )
+        );
+        write(mavenPath, updated);
+      }
+    }
+    catch (IOException e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Writes/overwrites metadata, replacing existing one, if any.
+   */
+  public void replace(final MavenPath mavenPath, final Maven2Metadata metadata) {
+    checkNotNull(mavenPath);
+    checkNotNull(metadata);
+    try {
+      write(mavenPath, toMetadata(metadata));
+    }
+    catch (IOException e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Deletes metadata.
+   */
+  public void delete(final MavenPath mavenPath) {
+    checkNotNull(mavenPath);
+    try {
+      mavenFacet.delete(mavenPath, mavenPath.hash(HashType.SHA1), mavenPath.hash(HashType.MD5));
+    }
+    catch (IOException e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  /**
+   * Converts NX VO into Apache Maven {@link Metadata}.
+   */
+  private Metadata toMetadata(final Maven2Metadata maven2Metadata) {
+    final Metadata result = new Metadata();
+    result.setModelVersion("1.1.0");
+    result.setGroupId(maven2Metadata.getGroupId());
+    result.setArtifactId(maven2Metadata.getArtifactId());
+    result.setVersion(maven2Metadata.getVersion());
+    if (maven2Metadata.getPlugins() != null) {
```

Comment:
mentioned already earlier, but each of these blocks would be simpler if the collections were always initialized and we didn't have to check for null